qi-lib/flow/core/deforest.rkt

```diff
@@ -0,0 +1,415 @@
+#lang racket/base
+
+(provide (for-syntax deforest-rewrite))
+
+(require (for-syntax racket/base
+                     syntax/parse
+                     racket/syntax-srcloc)
+         racket/performance-hint
+         racket/match
+         racket/list
+         racket/contract/base)
+
+;; These bindings are used for ~literal matching to introduce implicit
+;; producer/consumer when none is explicitly given in the flow.
+(define cstream->list #'-cstream->list)
+(define list->cstream #'-list->cstream)
+
+;; "Composes" higher-order functions inline by directly applying them
+;; to the result of each subsequent application, with the last argument
+;; being passed to the penultimate application as a (single) argument.
+;; This is specialized to our implementation of stream fusion in the
+;; arguments it expects and how it uses them.
+(define-syntax inline-compose1
+  (syntax-rules ()
+    [(_ f) f]
+    [(_ [op f] rest ...) (op f (inline-compose1 rest ...))]))
+
+(begin-for-syntax
+  ;; Partially reconstructs original flow expressions. The chirality
+  ;; is lost and the form is already normalized at this point though!
+  (define (prettify-flow-syntax stx)
+    (syntax-parse stx
+      #:datum-literals (#%host-expression esc #%blanket-template #%fine-template)
+      (((~literal thread)
```

Comment:
I'd like to know why this `thread` has to be literal; in the rest of the compiler it's matched by datum.